### ISSUE TITLE
Alltags view is extremely slow

### ISSUE THREAD

**Issue by leberrigan** [MotusDev/Motus-TO-DO#704]
I understand this is a well known issue with a workaround, but we keep getting collaborators who run into this issue and have to resort to the workaround. I wouldn't be surprised if the workaround was needed for most projects. 

**The workaround:** 
```
src <- tagme(projectID, update=T)

newview <- "CREATE VIEW alltags_fast as SELECT t1.hitID as hitID, t1.runID as runID, t1.batchID as batchID, t1.ts as ts, CASE WHEN t6.utcOffset is null then t1.ts else t1.ts - t6.utcOffset * 60 * 60 end as tsCorrected, t1.sig as sig, t1.sigSD as sigsd, t1.noise as noise, t1.freq as freq, t1.freqSD as freqsd, t1.slop as slop, t1.burstSlop as burstSlop, t2.done as done, CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end as motusTagID, t12.ambigID as ambigID, t2.ant as port, t2.len as runLen, t3.monoBN as bootnum, t4.projectID as tagProjID, t4.mfgID as mfgID, t4.type as tagType, t4.codeSet as codeSet, t4.manufacturer as mfg, t4.model as tagModel, t4.lifeSpan as tagLifespan, t4.nomFreq as nomFreq, t4.bi as tagBI, t4.pulseLen as pulseLen, t5.deployID as tagDeployID, t5.speciesID as speciesID, t5.markerNumber as markerNumber, t5.markerType as markerType, t5.tsStart as tagDeployStart, t5.tsEnd as tagDeployEnd, t5.latitude as tagDepLat, t5.longitude as tagDepLon, t5.elevation as tagDepAlt, t5.comments as tagDepComments, ifnull(t5.fullID, printf('?proj?-%d#%s:%.1f', t5.projectID, t4.mfgID, t4.bi)) as fullID, t3.motusDeviceID as deviceID, t6.deployID as recvDeployID, t6.latitude as recvDeployLat, t6.longitude as recvDeployLon, t6.elevation as recvDeployAlt, t6a.serno as recv, t6.name as recvDeployName, t6.siteName as recvSiteName, t6.isMobile as isRecvMobile, t6.projectID as recvProjID, t6.utcOffset as recvUtcOffset, t7.antennaType as antType, t7.bearing as antBearing, t7.heightMeters as antHeight, t8.english as speciesEN, t8.french as speciesFR, t8.scientific as speciesSci, t8.`group` as speciesGroup, t9.label as tagProjName, t10.label as recvProjName, NULL as gpsLat, NULL as gpsLon, NULL as gpsAlt FROM hits AS t1 LEFT JOIN runs AS t2 ON t1.runID = t2.runID left join allambigs t12 on t2.motusTagID = t12.ambigID LEFT JOIN batches AS t3 ON t3.batchID = t1.batchID LEFT JOIN tags AS t4 ON t4.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end LEFT JOIN tagDeps AS t5 ON t5.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end 	AND t5.tsStart = (SELECT max(t5b.tsStart) FROM tagDeps AS t5b WHERE t5b.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end 	AND t5b.tsStart <= t1.ts AND (t5b.tsEnd IS NULL OR t5b.tsEnd >= t1.ts) ) LEFT JOIN recvs as t6a on t6a.deviceID = t3.motusDeviceID LEFT JOIN recvDeps AS t6 ON t6.deviceID = t3.motusDeviceID AND t6.tsStart = (SELECT max(t6b.tsStart) FROM recvDeps AS t6b WHERE t6b.deviceID = t3.motusDeviceID 	AND t6b.tsStart <= t1.ts AND (t6b.tsEnd IS NULL OR t6b.tsEnd >= t1.ts) ) LEFT JOIN antDeps AS t7 ON t7.deployID = t6.deployID AND t7.port = t2.ant LEFT JOIN species AS t8 ON t8.id = t5.speciesID LEFT JOIN projs AS t9 ON t9.ID = t5.projectID LEFT JOIN projs AS t10 ON t10.ID = t6.projectID"
DBI::dbSendQuery(src$con, newview)

tbl.alltags <- tbl(src, "alltags_fast") # virtual table
```
**Reason for workaround:**
Caused by joining view with GPS table. Results in view taking 10's to 100's of times longer (hours to days) than before and often uses up all available memory.

**Solution**
We should look into getting this 'workaround' built in to the Motus R package so we don't need to keep directing folks to the script because honestly it's a bit sloppy and adds just another layer of complexity that we don't really need. The GPS table, while useful, is only necessary some of the time and may be best if we get collaborators to join it later on, after data has been filtered. This is how I did things in the past without an issue.

**Comment by adamdsmith** [comment]
I think this [was already addressed](https://github.com/MotusWTS/motus/commit/0af5ac76f00a4c52aa0f33b3a21d2e6cb8ce3f07), perhaps in `beta3`, by relegating the old `alltags` view with GPS info into a new `alltagsGPS` view, while the new `alltags` excludes GPS data. It does require either [some code to remove existing views](https://github.com/MotusDev/Motus-TO-DO/issues/421#issuecomment-565591275), or creating a new copy of the database.

**Comment by leberrigan** [comment]
Ahh, I see. So it just haven't been pushed to the master branch yet. I guess this raises a couple more points, then:
- Any timeline on pushing the update to the master branch?
- Is it possible to eliminate the need to create a new copy of the database with this update? We've made collaborators do this twice with the last two updates in recent months and it's made people quite frustrated as it takes a lot of time.

**Comment by adamdsmith** [comment]
I think the code I linked to remove some existing views, and then updating, will avoid the necessity of recreating the database.

I agree that it is quite annoying to have to recreate the database locally, whether it be for for `alltags` or to remove deprecated batches or whatever, and should be avoided as a solution unless absolutely necessary...

**Comment by denislepage** [comment]
In theory, that is relatively simple to apply to an existing database, particularly when it comes to views. Modifying table generally is more involved (copying table, dropping it, creating new format and re-inserting data). @steffilazerte  is unavailable for the next week or two, but can check on the status of the beta branch when she's back. We can easily create a function for now that will rewrite the view to work faster for those who need it until we have a cleaner solution as part of the package.

**Comment by leberrigan** [comment]
Just as an added note, Chapter 5.5 of the R book refers to the function '_filterByActivity_' which uses the '_alltags_' view so I had to modify it for one of our collaborators:
```

filterByActivity.fast <- function (src, return = "good", minLen = 3, maxLen = 5, 
          maxRuns = 100, ratio = 0.85) 
{
  if (!return %in% c("good", "bad", "all")) {
    stop("'return' must be one of 'good', 'bad', or 'all'", 
         call. = FALSE)
  }
  if (any(!is.numeric(c(minLen, maxLen, maxRuns, ratio)))) {
    stop("'minLen', 'maxLen', 'maxRuns', and 'ratio' must all be numeric", 
         call. = FALSE)
  }
  if (any(c(minLen, maxLen, maxRuns) < 1)) {
    stop("'minLen', 'maxLen', and 'maxRuns' must all be greater than 0", 
         call. = FALSE)
  }
  if (ratio < 0 | ratio > 1) 
    stop("'ratio' must be a value between 0 and 1", 
         call. = FALSE)
  if (minLen > maxLen) 
    stop("'minLen' must be smaller than or equal to 'maxLen'", 
         call. = FALSE)
  t <- DBI::dbListTables(src$con)
  if (any(!c("runs", "activity", "alltags") %in% 
          t)) {
    stop("'src' must contain at least tables 'activity', 'alltags', and 'runs'", 
         call. = FALSE)
  }
  tbl_runs <- dplyr::tbl(src$con, "runs") %>% dplyr::mutate(hourBin = round(.data$tsBegin/3600, 
                                                                            0))
  tbl_activity <- dplyr::tbl(src$con, "activity")
  if (nrow(DBI::dbGetQuery(src$con, "SELECT * FROM activity LIMIT 1")) < 
      1) {
    stop("'activity' table is empty, cannot filter by activity", 
         call. = FALSE)
  }
  tbl_bad <- dplyr::left_join(tbl_runs, tbl_activity, by = c(batchIDbegin = "batchID", 
                                                             "ant", "hourBin")) %>% dplyr::mutate(run2 = as.numeric(.data$run2), 
                                                                                                  numRuns = as.numeric(.data$numRuns)) %>% dplyr::filter(.data$len < 
                                                                                                                                                           maxLen) %>% dplyr::filter((.data$numRuns >= maxRuns & 
                                                                                                                                                                                        ((.data$run2/.data$numRuns) >= ratio)) | .data$len <= 
                                                                                                                                                                                       minLen) %>% dplyr::select("runID") %>% dplyr::mutate(probability = 0)
  tbl_prob <- dplyr::tbl(src$con, "alltags_fast") %>% dplyr::left_join(tbl_bad, 
                                                                  by = "runID") %>% dplyr::mutate(probability = dplyr::if_else(is.na(.data$probability), 
                                                                                                                               1, .data$probability))
  if (return == "good") 
    r <- dplyr::filter(tbl_prob, .data$probability > 0)
  if (return == "bad") 
    r <- dplyr::filter(tbl_prob, .data$probability == 0)
  if (return == "all") 
    r <- tbl_prob
  r
}
```

**Comment by steffilazerte** [comment]
This should be addressed in the new alltags view in motus v3.1.0
